### DISCUSSION TITLE
Property 'setPreviewData' does not exist on type 'NowResponse'

### DISCUSSION THREAD

**Discussion by Vadorequest** [vercel/next.js#13593] [Help] [answered]
# Bug report

## Describe the bug

Using `res.setPreviewData({});` throws:

`Property 'setPreviewData' does not exist on type 'NowResponse'.`

## To Reproduce

```ts
import { NowRequest, NowResponse } from '@now/node';
import { createLogger } from '@unly/utils-simple-logger';

import Sentry, { configureReq } from '../../utils/monitoring/sentry';

const fileLabel = 'api/preview';
const logger = createLogger({
  label: fileLabel,
});

export const error = async (req: NowRequest, res: NowResponse): Promise<void> => {
  try {
    configureReq(req);

    if (process.env.APP_STAGE !== 'production') {
      res.setPreviewData({}); // ERROR HERE

      console.log('Preview mode enabled'); // eslint-disable-line no-console
      res.writeHead(307, { Location: '/' });
      res.end();
    } else {
      res.end('');
      console.error('Preview mode is not allowed in production'); // eslint-disable-line no-console
    }
  } catch (e) {
    logger.error(e.message);

    Sentry.withScope((scope): void => {
      scope.setTag('fileLabel', fileLabel);
      Sentry.captureException(e);
    });

    res.json({
      error: true,
      message: process.env.APP_STAGE === 'production' ? undefined : e.message,
    });
  }
};

export default error;

```

## Expected behavior

It should work (type should exist).

## System information

- Version of `@now/node`: `1.6.1` (lastest)

**Comment by timneutkens** [comment, marked as answer]
`setPreviewData` can only be used in Next.js API routes (`pages/api/*`) not inside of Vercel API routes as this is a Next.js feature / integrated into Next.js. Potentially you're incorrectly typing your Next.js API route, it `NextApiRequest` and `NextApiRequest` from `'next'`:

```
import { NextApiRequest, NextApiRequest } from 'next'
```

**Reply by Vadorequest** [reply]
Thanks @timneutkens, indeed, I was using the wrong types from `@now/node`, I don't remember where that confusion came from, but I've been doing that for a while now.

I changed to the following:
![image](https://user-images.githubusercontent.com/3807458/83352478-f1091380-a34b-11ea-8f95-7823724e42b2.png)


```ts
import { NextApiRequest, NextApiResponse } from 'next';

type PreviewModeAPIQuery = {
  stop: string;
  redirectTo: string;
}

export const preview = async (req: NextApiRequest, res: NextApiResponse): Promise<void> => {
  try {
    configureReq(req);

    const {
      stop = 'false',
      redirectTo = '/',
    }: PreviewModeAPIQuery = req.query as PreviewModeAPIQuery;
...
```

**Reply by Vadorequest** [reply]
I went a bit deeper and actually completely removed all usage of `@now/node` in my Next.js app. Maybe they were from a time when Next.js didn't have proper types available?

Anyway, the correct way is to use the `NextApiRequest` and `NextApiResponse` as mentioned above.

Here is the PR, with mocking capabilities updated (to test API endpoints)
https://github.com/UnlyEd/next-right-now/pull/74/files
